[FAB-18117] Add policy checking to snapshot grpc service

wenjianqiao opened the pull request:
Signed-off-by: Wenjian Qiao wenjianq@gmail.com

#### Type of change
- New feature

#### Description
- Add ACLProvider to snapshot service to check ACL
- Add resources for snapshot submitrequest, cancelrequest and listpending
- Add CheckPolicyNoChannel to ACLProvider interface and implement CheckPolicyNoChannelBySignedData func in policy.go

#### Additional details
This PR is stacked on top of https://github.com/hyperledger/fabric/pull/1790

#### Related issues
https://jira.hyperledger.org/browse/FAB-18117

wenjianqiao committed:
[FAB-18117] Add policy checking to snapshot grpc service

- Add resources for snapshot submitrequest, cancelrequest and listpending
- Add CheckPolicyNoChannelBySignedData func to policy.go
- Add ACLProvider to snapshot service to check ACL

Signed-off-by: Wenjian Qiao <wenjianq@gmail.com>

ale-linux commented on core/ledger/snapshotgrpc/snapshot_service.go:
```diff
@@ -84,6 +103,33 @@ func (s *SnapshotService) QueryPendings(ctx context.Context, signedRequest *pb.S
+	}
+
+	expirationTime := crypto.ExpiresAt(signatureHdr.Creator)
+	if !expirationTime.IsZero() && time.Now().After(expirationTime) {
```
Just to double-check: the events for which the ACL is checked are never ordered events, right? In other words, there doesn't need to be consensus over the outcome of this check? Otherwise they would obviously lead to forks if time is not synchronised.

wenjianqiao replied:
@ale-linux Thanks for your comment. I confirm that the events for which the ACL is checked for snapshot service are never on orderer events or on commit path.

manish-sethi replied:
I did not notice this expiration check on other APIs that allow admin operations (e.g., join channel). Better not to have a different behavior for snapshot related APIs.

ale-linux replied:
+1 @manish-sethi, good point

wenjianqiao replied:
@manish-sethi The other APIs that allow admin ops are done via endorsor ProcessProposal that checks expiration via an auth filter expiration.go Therefore, individual cmd doesn't need separate call to check certificate expiration. However, because snapshot service is a separate grpc service, we should add this check.

manish-sethi replied:
I see, I missed that part here. Thanks for the clarification @wenjianqiao

manish-sethi replied:
Just to take a note, ideally, that should also have been performed at one place like ACLCheck but not related to this PR.

ale-linux commented on core/policy/policy.go:
```diff
@@ -172,3 +176,42 @@ func (p *policyChecker) CheckPolicyBySignedData(channelID, policyName string, sd
+
+// CheckPolicyNoChannelBySignedData checks that the passed signed data are valid with the respect to
+// passed policy on the local MSP.
+func (p *policyChecker) CheckPolicyNoChannelBySignedData(policyName string, signedData []*protoutil.SignedData) error {
```
Does this mean that all the elements in the `signedData` slice must comply with the policy?

wenjianqiao replied:
Yes, that's the intention.

ale-linux commented on core/ledger/snapshotgrpc/snapshot_service.go:
```diff
@@ -84,6 +103,33 @@ func (s *SnapshotService) QueryPendings(ctx context.Context, signedRequest *pb.S
+		return errors.New("missing signature header")
+	}
+
+	expirationTime := crypto.ExpiresAt(signatureHdr.Creator)
```
Just to double check: this ACL evaluation is not perfomed on the commit path right? Otherwise a time check like this would get us to fork.

wenjianqiao replied:
Same as above.

ale-linux commented on core/policy/policy.go:
```diff
@@ -144,7 +148,7 @@ func (p *policyChecker) CheckPolicyNoChannel(policyName string, signedProp *pb.S
 func (p *policyChecker) CheckPolicyBySignedData(channelID, policyName string, sd []*protoutil.SignedData) error {
 	if channelID == "" {
-		return errors.New("Invalid channel ID name during check policy on signed data. Name must be different from nil.")
+		return p.CheckPolicyNoChannelBySignedData(policyName, sd)
```
This function ends up being called in a ton of places, and this PR changes its failure model (prior to it, we would fail if the supplied channel ID was empty, now we don't). Have you verified that this change does not affect the verifier/committer path? Adding a new function to the 
```
type ACLProvider interface {
	//CheckACL checks the ACL for the resource for the channel using the
	//idinfo. idinfo is an object such as SignedProposal from which an
	//id can be extracted for testing against a policy
	CheckACL(resName string, channelID string, idinfo interface{}) error
}
```
interface (like a `CheckACLNoChannel` function) seems more desirable than changing the contract of the existing function, I'd go with that.

wenjianqiao replied:
When checking existing callers before making the change. I didn't find any code calling CheckACL with []*protoutil.SignedData. The existing code calls CheckACL with either SignedPropoal or Envelope (no []*protoutil.SignedData). However, you have a good point. It is better to add a CheckACLNoChannel and make the explicit call instead of overloading/changing existing behavior.

manish-sethi replied:
Even the existing behavior of the [policy checker](https://github.com/hyperledger/fabric/blob/545754c4bab3311ca56f6098f08dd43cec7c0cf5/core/policy/policy.go#L38) is inconsistent. Two similar functions [CheckPolicy](https://github.com/hyperledger/fabric/blob/545754c4bab3311ca56f6098f08dd43cec7c0cf5/core/policy/policy.go#L51) and [CheckPolicyBySignedData](https://github.com/hyperledger/fabric/blob/545754c4bab3311ca56f6098f08dd43cec7c0cf5/core/policy/policy.go#L145) behave differently, if no channelID is specified. 

@ale-linux  - I agree that adding the exposing the function `CheckACLNoChannel` at the `ACLProvider` level makes it clearer but I would add a note (I assume that you meant this implicitly as well) that we should modify the existing callers go via this function instead of first function above. Otherwise, it will simply add to the confusion looking at the code for - which would look like
- if you work with `SignedProposal` and have no channel call the function `CheckACL` with channelID as empty string
- if you work with `SignedData`, use the function `CheckACLNoChannel`.
This modification can be done in a separate PR and can open a Jira for this.

ale-linux replied:
Yes @manish-sethi, I share your view and the approach you propose

wenjianqiao replied:
@manish-sethi agree, I have thought the same and will open a Jira for it.

manish-sethi replied:
missed rolling back this change?

manish-sethi commented on core/ledger/snapshotgrpc/snapshot_service.go:
```diff
@@ -84,6 +103,33 @@ func (s *SnapshotService) QueryPendings(ctx context.Context, signedRequest *pb.S
+		Identity:  signatureHdr.Creator,
+		Data:      signedRequest.Request,
+		Signature: signedRequest.Signature,
+	}}
```
Nit: It's used immediately below, can be used directly in-line

manish-sethi commented on core/ledger/snapshotgrpc/snapshot_service_test.go:
```diff
@@ -87,37 +96,49 @@ func TestSnapshot(t *testing.T) {
+			channelID:     ledgerID,
+			signedRequest: createSignedQuery(ledgerID),
+			checkACLErr:   fmt.Errorf("fake-check-acl-error"),
+			errMsg:        "fake-check-acl-error",
```
I found this aclprovider error test a bit of a force fit into the existing test case - definition. I feel that having this as independent one makes a better sense.

wenjianqiao commented:
@ale-linux @manish-sethi I have addressed your comments. Also opened https://jira.hyperledger.org/browse/FAB-18185 to update the existing code to call CheckACLNoChannel when there is no channel.

ale-linux reviewed (commented):
LGTM. I won't merge yet to give @manish-sethi a chance to review as well.

manish-sethi commented on core/policy/policy.go:
```diff
@@ -144,7 +148,7 @@ func (p *policyChecker) CheckPolicyNoChannel(policyName string, signedProp *pb.S
 func (p *policyChecker) CheckPolicyBySignedData(channelID, policyName string, sd []*protoutil.SignedData) error {
 	if channelID == "" {
-		return errors.New("Invalid channel ID name during check policy on signed data. Name must be different from nil.")
+		return p.CheckPolicyNoChannelBySignedData(policyName, sd)
```
missed rolling back this change?

wenjianqiao replied:
Good catch. Done.

manish-sethi commented on core/policy/policy_test.go:
```diff
@@ -189,4 +189,24 @@ func TestPolicyChecker(t *testing.T) {
+		Signature: []byte("msg1"), // for testing only, signature is same as data to pass MockIdentity.Verify
+	}
+	identityDeserializer.Msg = signedData.Data
+	err = pc.CheckPolicyNoChannelBySignedData(mgmt.Admins, []*protoutil.SignedData{signedData})
```
- More of the error paths in this new function can be covered in the test. 
- It will be better to rearrange this test in the form of sub-test for the different APIs, so its easy to test the API. Not necessary to handle in this PR, as this was the existing pattern but better to improve if this does not take much to rearrange (will leave to you)

manish-sethi commented on core/aclmgmt/resourceprovider.go:
```diff
@@ -138,6 +141,11 @@ func (rp *aclmgmtPolicyProviderImpl) CheckACL(polName string, idinfo interface{}
+func (rp *aclmgmtPolicyProviderImpl) CheckACLNoChannel(resName string, idinfo interface{}) error {
+	return fmt.Errorf("overriding default acl provider is not supported")
+}
+
```
This really looks bad... an implementation knows the behavior of the consumer (which is usually the other way around). I am not well versed with this package but is it not possible for this impl to implement a separate interface which does not have this function - or rather does it have to implement this interface at all? I see this getting used only at one place in the production code where it does not need to implement this interface.

wenjianqiao replied:
Removed because peer resource does not allow override and never gets here.

manish-sethi commented on core/policy/policy_test.go:
```diff
@@ -162,31 +162,72 @@ func TestPolicyChecker(t *testing.T) {
+		require.Contains(t, err.Error(), "Failed evaluating policy on signed data during check policy on channel [C] with policy [readers]: [Invalid Signature]")
+	})
+
+	t.Run("CheckPolicyNoChannel", func(t *testing.T) {
+		sProp, _ := protoutil.MockSignedEndorserProposalOrPanic("A", &peer.ChaincodeSpec{}, []byte("Alice"), []byte("msg1"))
+		sProp.Signature = sProp.ProposalBytes
+
+		// Alice is a member of the local MSP, policy check must succeed
+		identityDeserializer.Msg = sProp.ProposalBytes
+		err := pc.CheckPolicyNoChannel(mgmt.Members, sProp)
+		require.NoError(t, err)
+
+		sProp, _ = protoutil.MockSignedEndorserProposalOrPanic("A", &peer.ChaincodeSpec{}, []byte("Bob"), []byte("msg2"))
+		// Bob is not a member of the local MSP, policy check must fail
+		err = pc.CheckPolicyNoChannel(mgmt.Members, sProp)
+		require.Error(t, err)
+		require.Contains(t, err.Error(), "Failed deserializing proposal creator during channelless check policy with policy [Members]: [Invalid Identity]")
+	})
+
+	t.Run("CheckPolicyNoChannel", func(t *testing.T) {
```
Nit: To differentiate, better to use different names, perhaps `CheckPolicyNoChannelSignedProposal` and `CheckPolicyNoChannelSignedData`. But this should not hold this PR from merging.
